diff --git a/kernel/compiler/compile.rb b/kernel/compiler/compile.rb
index eb5d181..3fdb352 100644
--- a/kernel/compiler/compile.rb
+++ b/kernel/compiler/compile.rb
@@ -33,11 +33,13 @@ class Compiler
   end
 
   def self.compile_file(path, flags=nil)
-    compiler.compile_file(path, flags)
+    cm = compiler.compile_file(path, flags)
+    VerifyYieldDebugger.run(cm)
   end
 
   def self.compile_string(string, context=nil, filename="(eval)", line=1)
-    compiler.compile_string(string, context, filename, line)
+    cm = compiler.compile_string(string, context, filename, line)
+    VerifyYieldDebugger.run(cm)
   end
 
   def self.execute(string)
diff --git a/kernel/compiler/compiled_file.rb b/kernel/compiler/compiled_file.rb
index 43ce2a4..c2cbdf0 100644
--- a/kernel/compiler/compiled_file.rb
+++ b/kernel/compiler/compiled_file.rb
@@ -1,3 +1,39 @@
+
+class VerifyYieldDebugger
+
+  def self.all_methods(obj)
+    case obj
+    when CompiledMethod
+      cmethods = [obj]
+      obj.literals.each do |o|
+        cmethods += all_methods(o)
+      end
+      cmethods
+    else
+      []
+    end
+  end
+
+  def self.check(cm)
+    iseq = cm.iseq.decode
+
+    iseq.each_with_index do |obj, idx|
+      if idx % 2 == 0 and obj != [:yield_debugger]
+        raise "iseq doesn't have enough yield_debugger instructions #{iseq.inspect}"
+      end
+    end
+  end
+
+  def self.run(cm_main)
+
+    for cm in all_methods(cm_main)
+      check(cm)
+    end
+
+    cm_main
+  end
+end
+
 module Rubinius
   ##
   # A decode for the .rbc file format.
@@ -34,6 +70,7 @@ module Rubinius
     ##
     # Writes the CompiledFile +cm+ to +file+.
     def self.dump(cm, file)
+      VerifyYieldDebugger.run(cm)
       File.open(file, "w") do |f|
         new("!RBIX", 0, "x").encode_to(f, cm)
       end
diff --git a/kernel/compiler/generator.rb b/kernel/compiler/generator.rb
index 23e05a1..e8238ef 100644
--- a/kernel/compiler/generator.rb
+++ b/kernel/compiler/generator.rb
@@ -220,6 +220,9 @@ class Compiler
 
       return unless what
 
+      @ip += 1
+      @stream << :yield_debugger
+
       @ip += what.size
       if what.size == 1
         @stream << what.first
@@ -230,6 +233,10 @@ class Compiler
 
     def flush
       return unless @last
+
+      @ip += 1
+      @stream << :yield_debugger
+
       @ip += @last.size
       if @last.size == 1
         @stream << @last.first
diff --git a/spec/frozen/core/io/reopen_spec.rb b/spec/frozen/core/io/reopen_spec.rb
index 1126ad1..60c5e0a 100644
--- a/spec/frozen/core/io/reopen_spec.rb
+++ b/spec/frozen/core/io/reopen_spec.rb
@@ -82,7 +82,7 @@ describe "IO#reopen" do
     @file1.gets.should == "Line 2: Two\n"
   end
 
-  platform_is_not :darwin, :freebsd do
+  platform_is_not :darwin, :freebsd, :linux do
     it "reassociates self with the I/O stream specified as an argument, after some sysreads" do
       length = 12 # length of first lines in numbered_lines.txt
 
diff --git a/vm/builtin/compiledmethod.cpp b/vm/builtin/compiledmethod.cpp
index db093fc..6cf4e5e 100644
--- a/vm/builtin/compiledmethod.cpp
+++ b/vm/builtin/compiledmethod.cpp
@@ -33,6 +33,7 @@ namespace rubinius {
     cm->local_count(state, Fixnum::from(0));
     cm->set_executor(CompiledMethod::default_executor);
     cm->backend_method_ = NULL;
+    cm->breakpoints = NULL;
 
     return cm;
   }
diff --git a/vm/builtin/compiledmethod.hpp b/vm/builtin/compiledmethod.hpp
index bf80025..4442914 100644
--- a/vm/builtin/compiledmethod.hpp
+++ b/vm/builtin/compiledmethod.hpp
@@ -37,7 +37,7 @@ namespace rubinius {
     /* accessors */
 
     VMMethod* backend_method_;
-
+    uint8_t* breakpoints;
 
     attr_accessor(name, Symbol);
     attr_accessor(iseq, InstructionSequence);
diff --git a/vm/builtin/system.cpp b/vm/builtin/system.cpp
index 495291a..c0bf9f4 100644
--- a/vm/builtin/system.cpp
+++ b/vm/builtin/system.cpp
@@ -19,6 +19,7 @@
 #include "objectmemory.hpp"
 #include "global_cache.hpp"
 #include "config_parser.hpp"
+#include "helpers.hpp"
 
 #include "builtin/array.hpp"
 #include "builtin/exception.hpp"
@@ -149,8 +150,9 @@ namespace rubinius {
 
   Fixnum* System::vm_fork(VM* state)
   {
+    pthread_mutex_lock(&rubinius::Helpers::mutex_yield_debugger);
     int result = ::fork();
-
+    pthread_mutex_unlock(&rubinius::Helpers::mutex_yield_debugger);
     if ( -1 == result ) {
       Exception::errno_error(state, "fork() failed!");
     }
diff --git a/vm/helpers.cpp b/vm/helpers.cpp
index 82b1e4f..25349b7 100644
--- a/vm/helpers.cpp
+++ b/vm/helpers.cpp
@@ -15,6 +15,7 @@
 #include "builtin/system.hpp"
 #include "builtin/thread.hpp"
 #include "builtin/channel.hpp"
+#include "builtin/iseq.hpp"
 
 #include "vm.hpp"
 #include "object_utils.hpp"
@@ -22,6 +23,22 @@
 
 namespace rubinius {
   namespace Helpers {
+
+    pthread_mutex_t mutex_yield_debugger = PTHREAD_MUTEX_INITIALIZER;
+    uint32_t num_debug_records = 0;
+    uint32_t debug_options = 0;
+    const char* debug_read_file = NULL;
+    const char* debug_write_file = NULL;
+
+    // breakpoint types
+    #define NEXT_BREAKPOINT 1
+    #define STEP_BREAKPOINT 2
+    #define CUSTOM_BREAKPOINT 4
+
+    // debug options
+    #define IGNORE_STEP_BREAKPOINT 1
+    #define IGNORE_CUSTOM_BREAKPOINT 2
+
     void add_method(STATE, CallFrame* call_frame, Module* mod, Symbol* name, CompiledMethod* method) {
       method->scope(state, call_frame->cm->scope());
       method->serial(state, Fixnum::from(0));
@@ -248,27 +265,302 @@ namespace rubinius {
       return module;
     }
 
+    void set_debug_files();
+    void write_debug_record(const char* out);
+    void write_debug_header(STATE, CallFrame* call_frame);
+    bool execute_debug_command(STATE, CallFrame* call_frame, const char* cmd);
+    void poll_debug_file(STATE, CallFrame* call_frame);
+    const char* extract_pid(const char* cmd, uintptr_t* pid);
+    void rtrim(char* str);
+    bool numeric_string(char* str);
+
     void yield_debugger(STATE, CallFrame* call_frame) {
-      Channel* chan;
+      CompiledMethod* cm = call_frame->cm;
+      int ip = call_frame->ip - 1;
+      uint32_t num_ops = cm->iseq()->opcodes()->num_fields();
+
+      // must not call any code that acquires the global_lock
+      // in the following section so there's no opportunity for deadlock
+
+      pthread_mutex_lock(&mutex_yield_debugger);
+
+      if(!cm->breakpoints) {
+        cm->breakpoints = new uint8_t[num_ops];   // optimize this to 3 bits
+        memset(cm->breakpoints, 0, num_ops);      // per opcode
+        cm->breakpoints[0] = STEP_BREAKPOINT;     /* never unset this */
+      }
+
+      if(cm->breakpoints[ip] != 0 &&
+            ((cm->breakpoints[ip] & NEXT_BREAKPOINT) ||
+            ((cm->breakpoints[ip] & STEP_BREAKPOINT) &&
+            !(debug_options & IGNORE_STEP_BREAKPOINT)) ||
+            ((cm->breakpoints[ip] & CUSTOM_BREAKPOINT) &&
+            !(debug_options & IGNORE_CUSTOM_BREAKPOINT)))) {
+
+        if(cm->breakpoints[ip] & NEXT_BREAKPOINT) {
+          cm->breakpoints[ip] ^= NEXT_BREAKPOINT;
+          if(debug_options & IGNORE_STEP_BREAKPOINT)
+            debug_options ^= IGNORE_STEP_BREAKPOINT;
+          if(debug_options & IGNORE_CUSTOM_BREAKPOINT)
+            debug_options ^= IGNORE_CUSTOM_BREAKPOINT;
+        }
+        else if(cm->breakpoints[ip] & STEP_BREAKPOINT) {
+          if(debug_options & IGNORE_CUSTOM_BREAKPOINT)
+            debug_options ^= IGNORE_CUSTOM_BREAKPOINT;
+        }
+        else if(cm->breakpoints[ip] & CUSTOM_BREAKPOINT) {
+          if(debug_options & IGNORE_STEP_BREAKPOINT)
+            debug_options ^= IGNORE_STEP_BREAKPOINT;
+        }
+
+        set_debug_files();
+        write_debug_header(state, call_frame);
+        poll_debug_file(state, call_frame);
+      }
+
+      pthread_mutex_unlock(&mutex_yield_debugger);
+    }
+
+    void set_debug_files() {
+      char* tmp;
+
+      if(!debug_read_file && !debug_write_file) {
+
+        tmp = getenv("RBX_DEBUG_DIR");
+
+        if(tmp) {
+          size_t len = strlen(tmp);
+          char* rfile = new char[len+25];
+          char* wfile = new char[len+25];
+
+          strcpy(rfile, tmp);
+          strcpy(wfile, tmp);
+
+          if(tmp[len - 1] != '/') {
+            rfile[len] = '/';
+            wfile[len] = '/';
+            ++len;
+          }
+          strcpy(rfile + len, "rbx_read.txt");
+          strcpy(wfile + len, "rbx_write.txt");
+          debug_read_file = rfile;
+          debug_write_file = wfile;
+        }
+      }
+    }
+
+    void write_debug_record(const char* out) {
+      FILE* wd;
+      char ssz[32];
+      std::string str;
+
+      if(!debug_write_file) return;
+      if((wd = fopen(debug_write_file, "a")) == NULL) return;
+
+      snprintf(ssz, 32 - 4, "%u\n", strlen(out));
+      str += ssz;
+      str += out;
+
+      fwrite(str.c_str(), 1, str.size(), wd);
+      fclose(wd);
+    }
+
+    void write_debug_header(STATE, CallFrame* call_frame) {
+      Object* self = call_frame->scope->self();
+      CompiledMethod* cm = call_frame->cm;
+      int ip = call_frame->ip - 1;
+      std::string str;
+      char tmp[2048];
+      Tuple* ops = cm->iseq()->opcodes();
+      opcode op = ops->at(state, ip + 1)->nil_p() ? 0 :
+          static_cast<Fixnum*>(ops->at(state, ip + 1))->to_native();
+
+      snprintf(tmp, 2048 - 4, "[debug] process id: %p, thread id: %p,\n"\
+                              "        self pointer: %p, method pointer: %p,\n"\
+                              "        ip: %d, instruction: '%s'\n",
+          (void*)getpid(), (void*)pthread_self(), self, cm, ip + 1,
+          InstructionSequence::get_instruction_name(op));
+      str += tmp;
+
+      if(kind_of<Class>(self) || kind_of<Module>(self)) {
+
+        if(static_cast<Module*>(self)->name() &&
+            !static_cast<Module*>(self)->name()->nil_p()) {
+          snprintf(tmp, 2048 - 4, "        self class name: '%s'\n",
+              static_cast<Module*>(self)->name()->c_str(state));
+          str += tmp;
+        }
+      }
+      else {
+        if(self->class_object(state)->name() &&
+            !self->class_object(state)->name()->nil_p()) {
+          snprintf(tmp, 2048 - 4, "        self class name: '%s'\n",
+              self->class_object(state)->name()->c_str(state));
+          str += tmp;
+        }
+      }
 
-      state->set_call_frame(call_frame);
+      if(cm->name() && !cm->name()->nil_p()) {
+        snprintf(tmp, 2048 - 4, "        method name: '%s'\n", cm->name()->c_str(state));
+        str += tmp;
+      }
+
+      if(!cm->backend_method_) {
+        snprintf(tmp, 2048 - 4, "        backend_method_ is NULL\n");
+        str += tmp;
+      }
+
+      write_debug_record(str.c_str());
+    }
+
+    void poll_debug_file(STATE, CallFrame* call_frame) {
+      FILE* rd;
+      uint32_t nth_record, sec_sleep = 2;
+      int sz_record;
+      char tmp[1024];
+      bool done = false;
+
+      if(!debug_read_file) return;
+
+      while(1) {
+
+        if((rd = fopen(debug_read_file, "r")) == NULL) {
+          sleep(sec_sleep);
+          continue;
+        }
+
+        nth_record = 0;
+
+        while(1) {
+
+          if(fgets(tmp, 1024 - 4, rd) == NULL) {
+            sleep(sec_sleep);
+            break;
+          }
+
+          rtrim(tmp);
+
+          if(!numeric_string(tmp)) {
+            sleep(sec_sleep);
+            break;
+          }
+
+          sz_record = atoi(tmp);
+          if(sz_record <= 0) continue;
+
+          if(sz_record > 1024 - 4) {
+            sleep(sec_sleep);
+            break;
+          }
+
+          tmp[sz_record] = '\0';
+          if(fread(tmp, 1, (size_t)sz_record, rd) < (size_t)sz_record) {
+            sleep(sec_sleep);
+            break;
+          }
 
-      chan = try_as<Channel>(G(vm)->get_ivar(state,
-            state->symbol("@debug_channel")));
+          if(nth_record >= num_debug_records) {
+            done = execute_debug_command(state, call_frame, tmp);
+            ++num_debug_records;
+          }
 
-      if(!chan) return;
+          ++nth_record;
+        }
+
+        fclose(rd);
+        if(done) return;
+      }
+    }
+
+    bool execute_debug_command(STATE, CallFrame* call_frame, const char* cmd) {
+      int ip = call_frame->ip - 1;
+      CompiledMethod* cm = call_frame->cm;
+      Tuple* ops = cm->iseq()->opcodes();
+      opcode op = ops->at(state, ip + 1)->nil_p() ? 0 :
+          static_cast<Fixnum*>(ops->at(state, ip + 1))->to_native();
+      uint32_t num_ops = ops->num_fields();
+      size_t ins_size = InstructionSequence::instruction_width(op);
+      uintptr_t pid;
+
+      cmd = extract_pid(cmd, &pid);
+      if(pid != 0 && pid != (uintptr_t)getpid()) return false;
+
+      if(strcmp(cmd, "next") == 0 || strcmp(cmd, "n") == 0) {
+        if(ip + ins_size + 1 < num_ops) {
+          cm->breakpoints[ip + ins_size + 1] |= NEXT_BREAKPOINT;
+          debug_options |= IGNORE_STEP_BREAKPOINT;
+        }
+        debug_options |= IGNORE_CUSTOM_BREAKPOINT;
+      }
+      else if(strcmp(cmd, "step") == 0 || strcmp(cmd, "s") == 0) {
+        debug_options |= IGNORE_CUSTOM_BREAKPOINT;
+      }
+      else if(strcmp(cmd, "run") == 0 || strcmp(cmd, "r") == 0) {
+        debug_options |= IGNORE_STEP_BREAKPOINT;
+      }
+      else {
+        write_debug_record("[debug] command not recognized\n");
+        return false;
+      }
 
-      Channel* control = state->thread->control_channel();
+      return true;
+    }
 
-      if(control->nil_p()) {
-        control = Channel::create(state);
-        state->thread->control_channel(state, control);
+    // expects the pid to be at the beginning of the string
+    // and to be hex characters (with or without a '0x' prefix).
+    // returns the start of the command after the pid
+    //
+    const char* extract_pid(const char* cmd, uintptr_t* pid) {
+      const char* p = cmd;
+      const char* strt = cmd;
+      const char* hex = cmd;
+      uintptr_t pow16 = 1;
+      *pid = 0;
+
+      while(*p) {
+        if(*p == ' ') {
+          hex = p - 1;
+          while(*p == ' ') ++p;
+          if(*p) strt = p;
+          break;
+        }
+        ++p;
       }
 
-      sassert(chan->has_readers_p());
+      if(hex >= cmd && *hex) {
+        do {
+          if(*hex >= '0' && *hex <= '9') {
+            *pid += (uintptr_t)(*hex - '0') * pow16;
+            pow16 *= 16;
+          }
+          else if(*hex >= 'a' && *hex <= 'f') {
+            *pid += (uintptr_t)(*hex - 'W') * pow16;
+            pow16 *= 16;
+          }
+          else if(*hex >= 'A' && *hex <= 'F') {
+            *pid += (uintptr_t)(*hex - '7') * pow16;
+            pow16 *= 16;
+          }
+          --hex;
+        } while(hex >= cmd);
+        return strt;
+      }
 
-      chan->send(state, state->thread.get());
-      control->receive(state, call_frame);
+      return cmd;
+    }
+
+    void rtrim(char* str) {   // don't use if strlen can equal 0
+      while(*str) ++str;
+      --str; *str = '\0';
+    }
+
+    bool numeric_string(char* str) {
+      if(!*str) return false;
+      while(*str) {
+        if(!isdigit(*str)) return false;
+        ++str;
+      }
+      return true;
     }
   }
 }
diff --git a/vm/helpers.hpp b/vm/helpers.hpp
index 5a0c3bf..52b94c2 100644
--- a/vm/helpers.hpp
+++ b/vm/helpers.hpp
@@ -1,6 +1,8 @@
 #ifndef RBX_HELPERS_HPP
 #define RBX_HELPERS_HPP
 
+#include <pthread.h>
+
 namespace rubinius {
   class CallFrame;
   class Module;
@@ -12,6 +14,9 @@ namespace rubinius {
   class LookupTableAssociation;
 
   namespace Helpers {
+
+    extern pthread_mutex_t mutex_yield_debugger;
+
     void add_method(VM*, CallFrame* call_frame, Module* mod, Symbol* name, CompiledMethod* meth);
     void attach_method(VM*, CallFrame* call_frame, Object* obj, Symbol* name, CompiledMethod* meth);
 
